IBX-11100: Hybrid tracking described in Dev-doc (#3250)
ezsystems/developer-documentation: docs/recommendations/raptor_integration/hybrid_tracking.md

|client       |yes           |Browser       |no                    |yes                                  |
|server       |no            |Server        |yes                   |yes (server side, on full page loads)|
|hybrid       |yes           |Server        |yes                   |yes                                  |

## Hybrid tracking flow

When hybrid tracking is enabled, `TrackingScriptExtension` renders the proxy bootstrap template and loads the `raptor-proxy.js` shim, which replaces the `window.raptor.push` and drains the initial event queue.
The shim sends events to the configured proxy endpoint, by default, `/raptor/track`.
Each event is sent in a separate request using a batch-compatible format.
The `TrackingProxyController::track` controller validates the `EventPayloadParser` payload, enriches each event with identifiers resolved from cookies, and dispatches one `TrackProxiedEventMessage` for every event.
Messages are then processed asynchronously via Messenger and `TrackProxiedEventMessageHandler` ultimately forwards them to Raptor through `TrackingService::trackRawParameters`.

### Hybrid tracking configuration

> To configure the Raptor hybrid tracking, use the `ibexa.system.<scope>.connector_raptor` configuration key in the `config/packages/ibexa_connector_raptor.yaml` file:

Reviewer: The intro points to `ibexa.system.<scope>.connector_raptor,` but the example sets hybrid_tracking_proxy_path at the top level under ibexa_connector_raptor. That's correct (it's a global setting, not per-scope), but easy to misread — could we add a short note that hybrid_tracking_proxy_path is configured globally, while enabled, customer_id, and tracking_type are per site access?

---
description: Enable hybrid tracking to avoid ad blockers and proxy events through your server.
month_change: true
---

# Raptor hybrid tracking

Hybrid tracking mode is an additional tracking mode available alongside [`client` and `server`](tracking_functions.md).

> In hybrid mode, the bundle includes a client-side shim that captures Raptor tracking events and sends them to a same-origin endpoint instead of communicating directly with Raptor SaaS.

Reviewer 2: I'm learning the "shim" world usage here.
It's correct. Maybe it's not simple English enough.

@mnocon Any feedback on "shim" usage? Is it well known? Does a definition with a link or in our text would worth it?

```suggestion
In hybrid mode, the bundle includes a client-side [shim](https://en.wikipedia.org/wiki/Shim_(computing)) that captures Raptor tracking events and sends them to a same-origin endpoint instead of communicating directly with Raptor SaaS.
```

Reviewer 3: Good idea with adding the link, it's not a commonly used expression

The server enriches each event with identifiers resolved from request cookies (`cookieId`, `sessionId`, and `userId`) and forwards it to Raptor asynchronously through [Ibexa Messenger](background_tasks.md).
Since the browser never connects to the Raptor domain, ad blockers cannot block the requests.

## Hybrid vs server or client-side tracking

Both `server` and `hybrid` tracking modes deliver pageviews and events server-side, so tracking requests are not affected by ad blockers.

> The main difference is that `hybrid` mode loads a local, first-party tracking JavaScript (`raptor-proxy.js`) provided by the DXP instance, instead of the Raptor SaaS JavaScript.

Reviewer 2: After reading all other definition, "local" might make more sense here at this point in the reading. Still, I would remove it to avoid possible misunderstood.

```suggestion
The main difference is that `hybrid` mode loads a first-party tracking JavaScript (`raptor-proxy.js`) provided by the DXP instance, instead of the Raptor SaaS JavaScript.
```

Event tracking still happens on the server, so ad blockers cannot block it.

The main advantage of `hybrid` mode over `server` mode is its ability to capture client side tracking events.
Visitor cookies are also refreshed more frequently than in `server` mode.

Check the table below to compare the behavior of different tracking modes:

|Tracking type|Browser script|Event delivery|Works with ad blockers|Cookie refresh                       |
|-------------|--------------|--------------|----------------------|-------------------------------------|
|client       |yes           |Browser       |no                    |yes                                  |
|server       |no            |Server        |yes                   |yes (server side, on full page loads)|
|hybrid       |yes           |Server        |yes                   |yes                                  |

## Hybrid tracking flow

> When hybrid tracking is enabled, `TrackingScriptExtension` renders the proxy bootstrap template and loads the `raptor-proxy.js` shim, which replaces the `window.raptor.push` function and drains the initial event queue.

Reviewer 3: Ok, so let's make it clearer:

```suggestion
When hybrid tracking is enabled, `TrackingScriptExtension` renders the proxy bootstrap template and loads the `raptor-proxy.js` shim, which replaces the `window.raptor.push` function and sends out the events queued before the tracking consent was given.
```

The shim sends each captured event to the proxy endpoint as a separate POST request.
The `TrackingProxyController::track` action validates the `EventPayloadParser` payload, enriches each event with identifiers resolved from cookies, and dispatches one `TrackProxiedEventMessage` for every event.
Messages are then processed asynchronously via Ibexa Messenger and `TrackProxiedEventMessageHandler` ultimately forwards them to Raptor through `TrackingService::trackRawParameters`.

### Hybrid tracking configuration

To configure the Raptor hybrid tracking, use the `ibexa.system.<scope>.connector_raptor` configuration key in the `config/packages/ibexa_connector_raptor.yaml` file:

``` yaml
ibexa_connector_raptor:

>     hybrid_tracking_proxy_path: '/track'

Reviewer 3: ```suggestion
    hybrid_tracking_proxy_path: '/raptor/track'
```
(same as https://github.com/ibexa/documentation-developer/pull/3250#discussion_r3492433621 )

ibexa:
    system:
        default:
            connector_raptor:
                enabled: true
                customer_id: '<id>'
                tracking_type: hybrid
```

!!! note

    The `hybrid_tracking_proxy_path` setting is configured globally, while `enabled`, `customer_id`, and `tracking_type` are configured per SiteAccess.

### Routing import

Hybrid tracking requires routing import.
It means that the proxy endpoint route must be imported in the project, for example in the `config/routes/ibexa_connector_raptor.yaml` file:

``` yaml
ibexa.connector_raptor:
    resource: '@IbexaConnectorRaptorBundle/Resources/config/routing.yaml'
```

This import is required only in `hybrid` mode.
Without it, the proxy endpoint returns 404 responses and tracking events are not processed.

### Proxy path configuration

You can configure the proxy endpoint path globally.
By default, it's set to `/raptor/track`.
The client side shim sends tracking events to this same-origin endpoint, which forwards them to Raptor asynchronously.

``` yaml
ibexa_connector_raptor:
    hybrid_tracking_proxy_path: '/track'
```

> !!! note

Reviewer 3: This note is broken in rendering

<img width="989" height="147" alt="Image" src="https://github.com/user-attachments/assets/171b3ebb-5151-4539-9f45-aa8a6994772d" />

---
description: Enable hybrid tracking to avoid ad blockers and proxy events through your server.
month_change: true
---

# Raptor hybrid tracking

Hybrid tracking mode is an additional tracking mode available alongside [`client` and `server`](tracking_functions.md).
In hybrid mode, the bundle includes a client side shim that captures Raptor tracking events and sends them to a same-origin endpoint instead of communicating directly with Raptor SaaS.

> The server enriches each event with identifiers resolved from request cookies (`cookieId`, `sessionId`, and `userId`) and forwards it to Raptor asynchronously through [Messenger](background_tasks.md).

Reviewer 3: ```suggestion
The server enriches each event with identifiers resolved from request cookies (`cookieId`, `sessionId`, and `userId`) and forwards it to Raptor asynchronously through [Ibexa Messenger](background_tasks.md).
```

Since the browser never connects to the Raptor domain, ad blockers cannot block the requests.

## Hybrid vs server or client side tracking

Both `server` and `hybrid` tracking modes deliver pageviews and events server side, so tracking requests are not affected by ad blockers.
The main difference is that `hybrid` mode loads a local, first-party tracking JavaScript (`raptor-proxy.js`) provided by the DXP instance, instead of the Raptor SaaS JavaScript.
The Raptor script itself (`//deliver.raptorstatic.com/script/raptor-3.0.min.js`) is loaded only in `client` mode.

> The browser script only forwards captured tracking events to the same-origin proxy endpoint on your DXP instance. First-party visitor cookies are created and refreshed server side, which is what helps them survive Safari ITP.

Reviewer 3: First time use of the acronym, please expand it - and I'd link to https://webkit.org/blog/7675/intelligent-tracking-prevention/ for people not familiar with this (I had to google)

Event tracking still happens on the server, so ad blockers cannot block it.

The main advantage of `hybrid` mode over `server` mode is its ability to capture client side tracking events.
Visitor cookies are also refreshed more frequently than in `server` mode.

Check the table below to compare the behavior of different tracking modes:

|Tracking type|Browser script|Event delivery|Works with ad blockers|Cookie refresh                       |
|-------------|--------------|--------------|----------------------|-------------------------------------|
|client       |yes           |Browser       |no                    |yes                                  |
|server       |no            |Server        |yes                   |yes (server side, on full page loads)|
|hybrid       |yes           |Server        |yes                   |yes                                  |

## Hybrid tracking flow

> When hybrid tracking is enabled, `TrackingScriptExtension` renders the proxy bootstrap template and loads the `raptor-proxy.js` shim, which replaces the `window.raptor.push` and drains the initial event queue.

Reviewer 3: ```suggestion
When hybrid tracking is enabled, `TrackingScriptExtension` renders the proxy bootstrap template and loads the `raptor-proxy.js` shim, which replaces the `window.raptor.push` function and drains the initial event queue.
```

I'm not sure what "drains the initial event queue". Isn't the initial event queue empty?

Reviewer: It's not empty — the bootstrap template already puts a pageview event in the queue when it sets up window.raptor. Events pushed before consent is given also stay in the queue, because the shim only loads after consent. So "drains the initial event queue" just means the shim sends out everything that was queued before it loaded (the pageview + anything else).

The shim sends events to the configured proxy endpoint, by default, `/raptor/track`.

> Each event is sent in a separate request using a batch-compatible format.

Reviewer 3: I'm not sure what it means. What's a batch-compatible format and why is it important to mention?

Reviewer: Fair point — it's an implementation detail that doesn't add much here. The request body wraps the event in an events array ({"events": [...]}), so the endpoint can accept multiple events per request, but the shim currently sends one event per request. The "batch-compatible format" wording isn't really useful for readers — it could just say "The shim sends each captured event to the proxy endpoint as a separate POST request."

> The `TrackingProxyController::track` controller validates the `EventPayloadParser` payload, enriches each event with identifiers resolved from cookies, and dispatches one `TrackProxiedEventMessage` for every event.

Reviewer 3: ```suggestion
The `TrackingProxyController::track` action validates the `EventPayloadParser` payload, enriches each event with identifiers resolved from cookies, and dispatches one `TrackProxiedEventMessage` for every event.
```
TrackingProxyController is a controller, but `track` is a action of that controller. Symfony also sometimes calls it "action method".

> Messages are then processed asynchronously via Messenger and `TrackProxiedEventMessageHandler` ultimately forwards them to Raptor through `TrackingService::trackRawParameters`.

Reviewer 3: ```suggestion
Messages are then processed asynchronously via Ibexa Messenger and `TrackProxiedEventMessageHandler` ultimately forwards them to Raptor through `TrackingService::trackRawParameters`.
```

### Hybrid tracking configuration

To configure the Raptor hybrid tracking, use the `ibexa.system.<scope>.connector_raptor` configuration key in the `config/packages/ibexa_connector_raptor.yaml` file:

``` yaml
ibexa_connector_raptor:
    hybrid_tracking_proxy_path: '/track'

ibexa:
    system:
        default:
            connector_raptor:
                enabled: true
                customer_id: '<id>'
                tracking_type: hybrid
```

!!! note

    The `hybrid_tracking_proxy_path` setting is configured globally, while `enabled`, `customer_id`, and `tracking_type` are configured per SiteAccess.

### Routing import

Hybrid tracking requires routing import.

> It means that the proxy endpoint route must be imported in the project, for example in the `config/routes.yaml` file:

Reviewer 3: ```suggestion
It means that the proxy endpoint route must be imported in the project, for example in the `config/routes/ibexa_connector_raptor.yaml` file:
```

This is where the import is added by default by recipes (https://github.com/ibexa/recipes-dev/pull/250/changes), let's use this so that people who have it added by recipes don't add it twice

``` yaml
ibexa.connector_raptor:
    resource: '@IbexaConnectorRaptorBundle/Resources/config/routing.yaml'
```

This import is required only in `hybrid` mode.
Without it, the proxy endpoint returns 404 responses and tracking events are not processed.

### Proxy path configuration

You can configure the proxy endpoint path globally.
By default, it's set to `/raptor/track`.
The client side shim sends tracking events to this same-origin endpoint, which forwards them to Raptor asynchronously.

``` yaml
ibexa_connector_raptor:
    hybrid_tracking_proxy_path: '/track'
```

!!! note

   The path must start with a `/` symbol and cannot overlap with any existing routes.
   The route is generated from this value, so updating it automatically updates the route as well.
   You don't need to modify the routing import.

After changing the configuration, clear the application cache:

``` bash
php bin/console cache:clear
```

Verify that the route has been registered correctly.
To do it, run the following command:

``` bash
php bin/console debug:router | grep raptor
```

### Template usage

By default, the tracking script waits for user consent before sending tracking events.

> Consent can be provided either by rendering the script with `ibexa_tracking_script` (`has_consented` = `true`) or by triggering the `enableTracking` JavaScript event.

Reviewer 3: I feel this part is repeated in https://ez-systems-developer-documentation--3250.com.readthedocs.build/en/3250/templating/twig_function_reference/recommendations_twig_functions/#ibexa_tracking_script-function:~:text=needed%2E-,If,consent,-%2E , but here it's less detailed, doesn't mention `enableTracking` and doesn't show an example.

Can we just say something like:
```
By default, the tracking script waits for user consent before sending tracking events. For more info, see https://ez-systems-developer-documentation--3250.com.readthedocs.build/en/3250/templating/twig_function_reference/recommendations_twig_functions/#ibexa_tracking_script-function
```

And maybe in recommendations_twig_functions add a new subsection called "Handle tracking consent" for easier linking between sections?

Reviewer: Agreed — no point duplicating it here, especially since this version is less complete (no enableTracking, no example). Shortening it to a one-line note plus a link to ibexa_tracking_script makes sense, and adding a "Handle tracking consent" subsection there for easy linking is a good idea.

> The same template syntax is used across all tracking modes, so changing `tracking_type` does not require any template changes.

Reviewer 3: ```suggestion
The same template syntax is used across all tracking modes, so changing `tracking_type` doesn't require any template changes.
```
